Plugins/Tweaks/Tweaks/FixDispelEffectLevels.cpp

```diff
@@ -0,0 +1,166 @@
+#include "Tweaks/FixDispelEffectLevels.hpp"
+
+#include "Services/Hooks/Hooks.hpp"
+#include "Utils.hpp"
+
+#include "API/CAppManager.hpp"
+#include "API/CNWCCMessageData.hpp"
+#include "API/CNWSCreature.hpp"
+#include "API/CNWSCreatureStats.hpp"
+#include "API/CNWSObject.hpp"
+#include "API/CNWSPlayer.hpp"
+#include "API/CNWSRules.hpp"
+#include "API/CServerAIMaster.hpp"
+#include "API/CServerExoApp.hpp"
+#include "API/Functions.hpp"
+#include "API/Globals.hpp"
+
+namespace Tweaks
+{
+
+using namespace NWNXLib;
+using namespace NWNXLib::API;
+
+FixDispelEffectLevels::FixDispelEffectLevels(Services::HooksProxy* hooker)
+{
+    hooker->RequestExclusiveHook<Functions::_ZN21CNWSEffectListHandler21OnApplyDispelAllMagicEP10CNWSObjectP11CGameEffecti>
+        (&CNWSEffectListHandler__OnApplyDispelAllMagic);
+    hooker->RequestExclusiveHook<Functions::_ZN21CNWSEffectListHandler22OnApplyDispelBestMagicEP10CNWSObjectP11CGameEffecti>
+        (&CNWSEffectListHandler__OnApplyDispelBestMagic);
+}
+
+int32_t FixDispelEffectLevels::CNWSEffectListHandler__OnApplyDispelAllMagic(CNWSEffectListHandler* thisPtr, CNWSObject* pObject, CGameEffect* pEffect, BOOL bLoadingGame)
+{
+    (void)thisPtr;
+    (void)bLoadingGame;
+    auto nDispelLevel = pEffect->GetInteger(0);
+    int nDispelledEffects = 0;
+    auto* pCreatorObject = Utils::GetGameObject(pEffect->m_oidCreator);
+    if (nDispelLevel < 1)
+    {
+        if (auto* pCreature = Utils::AsNWSCreature(pCreatorObject))
+            nDispelLevel = pCreature->m_pStats->GetLevel(0);
+        else
+            return 1;
+    }
+
+    std::vector<uint64_t> nIgnoreEffectIDs;
+    auto* pMessageData = new CNWCCMessageData();
```

Comment:
Don't use `new` for objects unless NWN has to take ownership of it. Either `unique_ptr<CNWCCMessageData>`, or allocate it on the stack directly. That way you can't forget to do a delete. Stack alloc is also marginally faster.